core/trino-main/src/main/java/io/trino/execution/scheduler/policy/PrioritizeUtilizationExecutionSchedule.java

```diff
@@ -0,0 +1,559 @@
+                nonLazyFragments.add(fragment.getId());
+            }
+            return new FragmentSubGraph(upstreamFragments, lazyUpstreamFragments, false);
+        }
+
+        @Override
+        public FragmentSubGraph visitJoin(JoinNode node, PlanFragmentId currentFragmentId)
+        {
+            return processJoin(
+                    node.getDistributionType().orElseThrow() == JoinNode.DistributionType.REPLICATED,
+                    node.getLeft(),
+                    node.getRight(),
+                    currentFragmentId);
+        }
+
+        @Override
+        public FragmentSubGraph visitSpatialJoin(SpatialJoinNode node, PlanFragmentId currentFragmentId)
+        {
+            return processJoin(
+                    node.getDistributionType() == SpatialJoinNode.DistributionType.REPLICATED,
+                    node.getLeft(),
+                    node.getRight(),
+                    currentFragmentId);
+        }
+
+        @Override
+        public FragmentSubGraph visitSemiJoin(SemiJoinNode node, PlanFragmentId currentFragmentId)
+        {
+            return processJoin(
+                    node.getDistributionType().orElseThrow() == SemiJoinNode.DistributionType.REPLICATED,
+                    node.getSource(),
+                    node.getFilteringSource(),
+                    currentFragmentId);
+        }
+
+        @Override
+        public FragmentSubGraph visitIndexJoin(IndexJoinNode node, PlanFragmentId currentFragmentId)
+        {
+            return processJoin(
+                    true,
+                    node.getProbeSource(),
+                    node.getIndexSource(),
+                    currentFragmentId);
+        }
+
+        private FragmentSubGraph processJoin(boolean replicated, PlanNode probe, PlanNode build, PlanFragmentId currentFragmentId)
+        {
+            FragmentSubGraph probeSubGraph = probe.accept(this, currentFragmentId);
+            FragmentSubGraph buildSubGraph = build.accept(this, currentFragmentId);
+
+            // start probe source stages after all build source stages finish
+            addDependencyEdges(buildSubGraph.getUpstreamFragments(), probeSubGraph.getLazyUpstreamFragments());
+
+            boolean currentFragmentLazy = probeSubGraph.isCurrentFragmentLazy() && buildSubGraph.isCurrentFragmentLazy();
+            if (replicated && currentFragmentLazy) {
+                // Do not start join stage (which can also be a source stage with table scans)
+                // for replicated join until build source stage enters FLUSHING state.
+                // Broadcast join limit for CBO is set in such a way that build source data should
+                // fit into task output buffer.
+                // In case build source stage is blocked on full task buffer then join stage
+                // will be started automatically regardless od dependency. This is handled by
+                // unblockStagesWithFullOutputBuffer method.
```

Comment:
please make sure it is test covered. Prefereably with an integration test.

Reply:
It is tested in `TestPrioritizeUtilizationExecutionSchedule#testBroadcastSourceJoin`:
```
        // join stage should start after build stage buffer is full
        buildStage.setAnyTaskBlocked(true);
        schedule.schedule();
        assertThat(getActiveFragments(schedule)).containsExactlyInAnyOrder(joinSourceFragment.getId(), buildFragment.getId());
```